### PR TITLE
Launch workflows against multiple sets [SATURN-1003]

### PR DESCRIPTION
**Pull request by zarsky-broad**
This handles combining sets when the singular entity is the root type, as well as launching against multiple sets in parallel when the set is the root type.

Desired behaviors were figured out working together with Sushma.

### CODE REVIEW

**Comment by breilly2** [comment]
Initial testing note: I noticed that items in the sets are not de-duped. With a singular root entity type, I chose two sets that both had the same two elements. It happily created a new set with 4 items, each of the two listed twice. It looks like Rawls _did_ de-dupe it and only launched 2 workflows.

I'm kinda thinking that we should recognize this and... do something different than what it's currently doing. If the sets don't entirely overlap, then creating a new set with the unique items makes sense. If the sets do entirely overlap, we could create a new set, though that would result in a 3rd copy of the same set. Might be most helpful to let the user know what's going on, but that's a very specific and uniquely different experience for a (likely) very uncommon corner case. Other thoughts?

**Comment by breilly2** [comment]
Screen Shot 2019-09-13 at 10 04 30 AM
> This will launch 2 analyses simultaneously.

Nice!

File: src/pages/workspaces/workspace/workflows/DataStepContent.js [hunk 1]
```diff
@@ -41,10 +41,11 @@ export default class DataStepContent extends Component {
   isValidSelectionModel() {
     const { entitySelectionModel } = this.state
     const { newSetName, selectedEntities, type } = entitySelectionModel
-    const { entityType, name } = selectedEntities
+    const selectionSize = _.size(selectedEntities)
+
     return (type === EntitySelectionType.processAll ||
-      (type === EntitySelectionType.processFromSet && !!entityType && !!name) ||
-      (_.size(selectedEntities) > 0 && !!newSetName))
+      (selectionSize > 0 && !!newSetName && (type !== EntitySelectionType.chooseSet || selectionSize <= 10))
```
Comment (breilly2) [inline src/pages/workspaces/workspace/workflows/DataStepContent.js]: You faithfully and correctly simplified this conditional. I now realize that the conditional as I originally wrote it was a little sloppy with requiring `newSetName` even if only a single row is selected. That tripped me up when trying to read and understand the intent of this change.

I decided to see what it would look like to explicitly write this longhand without de-duping any of the conditionals across selection modes.

```suggestion
      (type === EntitySelectionType.chooseRows && (selectionSize === 1 || (selectionSize > 1 && !!newSetName))) ||
      (type === EntitySelectionType.chooseSet && (selectionSize === 1 || (selectionSize > 1 && selectionSize <= 10 && !!newSetName)))
```

What do you think?

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js [hunk 2]
```diff
@@ -88,10 +108,20 @@ export default ajaxCaller(class LaunchAnalysisModal extends Component {
         this.createSetAndLaunch(entities)
       }
     } else if (type === EntitySelectionType.processFromSet) {
-      const { entityType, name } = selectedEntities
-      this.launch(entityType, name, `this.${rootEntityType}s`)
+      if (_.size(selectedEntities) === 1) {
+        const { entityType, name } = selectedEntities[0]
+        this.launch(entityType, name, `this.${rootEntityType}s`)
+      } else {
+        const entities = _.reduce((acc, { attributes: { [`${rootEntityType}s`]: { items } } }) => _.concat(acc, _.map('entityName', items)), [],
```
Comment (breilly2) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: TOL: I finally figured out what this is doing. Can it be done with `_.flatMap`? Or can it `_.flow` somehow?

File: src/pages/workspaces/workspace/workflows/WorkflowView.js [hunk 3]
```diff
@@ -506,16 +507,15 @@ const WorkflowView = _.flow(
 
   describeSelectionModel() {
     const { modifiedConfig: { rootEntityType }, entityMetadata, entitySelectionModel: { newSetName, selectedEntities, type } } = this.state
-    const { name } = selectedEntities
     const count = _.size(selectedEntities)
     const newSetMessage = count > 1 ? `(will create a new set named "${newSetName}")` : ''
```
Comment (breilly2) [inline src/pages/workspaces/workspace/workflows/WorkflowView.js]:
```suggestion
     const newSetMessage = (type === EntitySelectionType.processAll || count > 1) ? `(will create a new set named "${newSetName}")` : ''
```
Otherwise `newSetName` will be empty in the process-all case.

**Review by breilly2 — COMMENTED** [review]
Mid-way review. I've reviewed all of the UI. I have not yet reviewed the addition of `launchParallel` and related changes to `baseLaunch`, etc.

File: src/pages/workspaces/workspace/workflows/DataStepContent.js [hunk 4]
```diff
@@ -41,10 +41,12 @@ export default class DataStepContent extends Component {
...
     const { entitySelectionModel } = this.state
     const { newSetName, selectedEntities, type } = entitySelectionModel
-    const { entityType, name } = selectedEntities
+    const selectionSize = _.size(selectedEntities)
+
     return (type === EntitySelectionType.processAll ||
-      (type === EntitySelectionType.processFromSet && !!entityType && !!name) ||
-      (_.size(selectedEntities) > 0 && !!newSetName))
+      (type === EntitySelectionType.chooseRows && (selectionSize === 1 || (selectionSize > 1 && !!newSetName))) ||
+      (type === EntitySelectionType.chooseSet && (selectionSize === 1 || (selectionSize > 1 && selectionSize <= 10 && !!newSetName)))
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/DataStepContent.js]: Currently the `processFromSet` is always invalid. There should be a case for that, which I think should validate `newSetName` instead of this line.

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js [hunk 5]
```diff
@@ -88,10 +108,20 @@ export default ajaxCaller(class LaunchAnalysisModal extends Component {
...
+          selectedEntities)
+        this.createSetAndLaunch(entities)
+      }
     } else if (type === EntitySelectionType.chooseSet) {
-      this.launch(rootEntityType, selectedEntities['name'])
+      if (_.size(selectedEntities) === 1) {
+        this.launch(rootEntityType, selectedEntities['name'])
+      } else {
+        this.launchParallel()
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js:128]: TOL: Going forward we could consider making multi-launch the general case, with single-launch as a specific case of that. That would allow us to leverage the nice status and error handling for both.

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js [hunk 6]
```diff
@@ -127,23 +157,59 @@ export default ajaxCaller(class LaunchAnalysisModal extends Component {
...
+    const { onSuccessMulti, entitySelectionModel: { selectedEntities }, config: { rootEntityType } } = this.props
+
+    this.setState({ multiLaunchCompletions: 0, message: undefined })
+
+    const allErrors = await Promise.all(_.map(async ({ name }) => {
+      try {
+        await this.baseLaunch(rootEntityType, name)
+        this.setState(_.update('multiLaunchCompletions', _.add(1)))
+      } catch (error) {
+        this.setState(_.update('multiLaunchCompletions', _.add(1)))
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: nit: This could be in a `finally` to avoid having it twice.

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js [hunk 7]
```diff
@@ -51,11 +56,26 @@ export default ajaxCaller(class LaunchAnalysisModal extends Component {
         }, ['Launch']) :
         h(ButtonPrimary, { onClick: onDismiss }, ['OK'])
     }, [
-      div({ style: { margin: '1rem 0' } }, [
-        'This analysis will be run by ', h(CromwellVersionLink), '.'
+      div({ style: { margin: '1rem 0' } }, ['This analysis will be run by ', h(CromwellVersionLink), '.']),
+      type === EntitySelectionType.chooseSet && div({ style: { margin: '1rem 0' } }, [`This will launch ${entityCount} analyses simultaneously.`]),
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: Should this be conditional on having multiple sets selected?

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js (hunk 6 above)
Comment (zarsky-broad) [reply src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: Ah, the duplication was left over from an old idea, good catch

File: src/pages/workspaces/workspace/workflows/DataStepContent.js (hunk 4 above)
Comment (zarsky-broad) [reply src/pages/workspaces/workspace/workflows/DataStepContent.js]: I _think_ I fixed this?

File: src/pages/workspaces/workspace/workflows/EntitySelectionType.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
-export default Object.freeze({
-  processAll: 'process all',
-  processFromSet: 'choose existing',
-  chooseRows: 'choose rows',
-  chooseSet: 'choose set'
-})
+export default {
+  processAll: Symbol('process all'),
+  processFromSet: Symbol('process from set'),
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/EntitySelectionType.js]: Thanks for making this match up, it helps with debugging!

File: src/pages/workspaces/workspace/workflows/DataStepContent.js [hunk 9]
```diff
@@ -41,10 +43,10 @@ export default class DataStepContent extends Component {
...
     const { entitySelectionModel } = this.state
     const { newSetName, selectedEntities, type } = entitySelectionModel
-    const { entityType, name } = selectedEntities
-    return (type === EntitySelectionType.processAll ||
-      (type === EntitySelectionType.processFromSet && !!entityType && !!name) ||
-      (_.size(selectedEntities) > 0 && !!newSetName))
+    const selectionSize = _.size(selectedEntities)
+
+    return selectionSize === 1 ||
+      (!!newSetName && (type === processAll || (selectionSize > 1 && (type === chooseRows || selectionSize <= 10))))
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/DataStepContent.js]: nit: technically don't need to validate `newSetName` in the `chooseSet` case, but since it gets initialized to a value it doesn't seem to cause problems.

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js [hunk 10]
```diff
@@ -127,23 +162,59 @@ export default ajaxCaller(class LaunchAnalysisModal extends Component {
...
+      } finally {
+        this.setState(_.update('multiLaunchCompletions', _.add(1)))
+      }
+    }, selectedEntities))
+
+    const multiLaunchErrors = _.compact(allErrors)
+
+    if (_.isEmpty(multiLaunchErrors)) {
+      onSuccessMulti()
+    } else if (multiLaunchErrors.length === selectedEntities.length) {
```
Comment (panentheos) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: This needs to use `_.size` for `selectedEntities` since it's an object.

**Review by panentheos — APPROVED** [review]
Looks good, see one comment.

File: src/pages/workspaces/workspace/workflows/DataStepContent.js (hunk 9 above)
Comment (breilly2) [reply src/pages/workspaces/workspace/workflows/DataStepContent.js]: ➕ More than a nit to me. I want to make sure we leave this code in a state where it is obviously doing the right thing. If it looks like there's a case that isn't fully considered or is making needless checks, then we won't know what the correct behavior is supposed to me when we come back to this code a year from now... and the corner cases are too complex for us to reliably always test all of them manually when making changes.
